### ISSUE TITLE
st-sideload-relay: Fetch incoming binlogs into the relay log path instead of /tmp

### ISSUE THREAD

**Issue by gordan-bobic** [shatteredsilicon/silicon-toolkit#6]
Because /tmp is almost certainly on a different file system, when we copy GBs of files from /tmp to the relay log path, this is much more expensive than a local file rename on the same file system.

Fetch binlogs into temporary files in the same directory where the relay logs are, called `.stsr-$replication_connection_name-$upstream_binlog_name` (add `.stsr-$replication_connection_name-` prefix to avoid collissions. Then rename them into relay logs as before.

**Comment by oblitorum** [comment]
Actually, I added a `--tmp-dir` argument for specifying where to store those temporary binlogs, although the default value for this is `/tmp` currently. I think maybe we could keep this argument and change the default value to the relay log path.

**Comment by oblitorum** [comment]
Can't think of a good way to achieve this without using a file/database to store those file records. Although I learnt that there is a feature in Unix called 'anonymous temporary file', which seems a good way to avoid problems with stale files, but I'm not very sure if it fits in with mysqlbinlog, might worth to do some tests.

**Comment by gordan-bobic** [comment]
mysqlbinlog fetches files one by one, right?

Hows' this for an idea for an algorithm outline:

1) Get the next binlog name from upstream
2) Prefix with `.stsr-$replication_connection_name-`
Default for $replication_connection_name is empty, so in default case it would be just a `.stsr-` prefix.
3) Check if the constructed file name already exists in the target path. If it does - exit with an error that suggests specifying a directory other than the directory where the relay logs are using `--tmp-dir` option.
4) If the constructed file name doesn't already exists, fetch into that file name
5) As soon as the download is complete, rename and add to the relay list and refresh by restarting the slave thread.
6) If we are not over the maximum relay size or io-thread-threshold, or lag is less than --min-lag GOTO 1), otherwise sleep for 10 seconds and repeat 6)

Am I missing an edge case here? Or would that work reasonably reliably?

**Comment by oblitorum** [comment]
> Check if the constructed file name already exists in the target path. If it does - exit with an error that suggests specifying a directory other than the directory where the relay logs are using --tmp-dir option.

I think we should double check that this filename exists in `SHOW BINARY LOGS` or file `relay-log.index`, assume the `st-sideload-relay` progress died/killed right after the `.stsr-$replication_connection_name-` file is fetched, it will also leave a file with constructed file name.

Thinking of this, maybe we could remove stale files if we find they are not in `SHOW BINARY LOGS` or file `relay-log.index`, but this will not work if `st-sideload-relay` never runs again.

**Comment by gordan-bobic** [comment]
Yes, we should abort if the target file exists, e.g. if there was a sudden crash or somebody did `kill -9` on `st-sideload-relay`. But cases like this should be very rare.

Thinking about it, the one edge case would be where two instances of st-sideload-relay end up running at the same time. I'm thinking a lock file in /var/lock/stsr-$replication_connection_name.lock, overrideable with `--lock-file` parameter?

That would still potentially leave us open to a collission if something like mysqld_multi is used, but if somebody got as far as using something as obscure as mysqld_multi, they can probably be trusted to cover that edge case themselves.

**Comment by gordan-bobic** [comment]
Just to clarify - we should still rename the incoming file into the relay log name pattern. This is purely about temporary incoming binlogs being retrieved into the same directory where the relay logs already are to avoid cross-filesystem mv because mv of gigabytes of binlogs between file systems is slow. Just in case I didn't explain it properly in the original post.

**Comment by oblitorum** [comment]
OK, got it
